log/logr/packetlogger.go

```diff
@@ -155,3 +155,48 @@ func errLogsToStderr(c zap.Config) zap.Option {
 	})
 	return splitLogger
 }
+
+// handleFields converts a bunch of arbitrary key-value pairs into Zap fields.  It takes
+// additional pre-converted Zap fields, for use with automatically attached fields, like
+// `error`.
+func handleFields(l *zap.Logger, args []interface{}, additional ...zap.Field) []zap.Field {
+	// a slightly modified version of zap.SugaredLogger.sweetenFields
+	if len(args) == 0 {
+		// fast-return if we have no suggared fields.
+		return additional
+	}
+
+	// unlike Zap, we can be pretty sure users aren't passing structured
+	// fields (since logr has no concept of that), so guess that we need a
+	// little less space.
+	fields := make([]zap.Field, 0, len(args)/2+len(additional))
+	for i := 0; i < len(args); {
+		// check just in case for strongly-typed Zap fields, which is illegal (since
+		// it breaks implementation agnosticism), so we can give a better error message.
+		if _, ok := args[i].(zap.Field); ok {
+			l.DPanic("strongly-typed Zap Field passed to logr", zap.Any("zap field", args[i]))
+			break
+		}
+
+		// make sure this isn't a mismatched key
+		if i == len(args)-1 {
+			l.DPanic("odd number of arguments passed as key-value pairs for logging", zap.Any("ignored key", args[i]))
+			break
+		}
```

Comment:
can't we just do this check out side of the loop, `if len(args) % 2 != 0 {`

Reply:
if this is outside the loop then no pairs get added, with it where it is only a final key without a value will be omitted.

Reply:
oh right the DPanic only panics when debugging but will continue if not debugging. I thought it would always panic.